heap.go

```diff
@@ -22,16 +22,21 @@ type Heap struct {
 
 */
 
-/* TODO: shouldn't only build up on the left hand side */
 func insert(heap *Heap, val int) *Heap {
+  insertion := val
   if heap == nil {
     heap = &Heap{ nil, val, nil }
+    return heap
   } else if val > heap.value {
-    heap.left = insert(heap.left, val)
-  } else if val < heap.value {
-    heap.left = insert(heap.left, heap.value)
+    insertion = heap.value
     heap.value = val
   }
+  // not sure about the costs of these 'assignments' when already assigned
```

Comment:
I'm sure this can be done in a simpler way. It is an improvement still so I'll just merge and raise an issue about it.